test/test_optim.py

```diff
@@ -2199,6 +2200,202 @@ def test_non_empty_state(self, device, dtype, optim_info):
             for state in optim.state.values():
                 self.assertGreater(len(state), 0)
 
+    @optims(
+        [optim for optim in optim_db if optim.optim_cls.__name__ == "Adam"],
+        dtypes=[torch.float32],
+    )
+    def test_less_mem_beta1_zero_adam(self, device, dtype, optim_info):
+        # test that beta1=0.0 uses less memory than the default
+        model = torch.nn.Linear(5, 5)
+        model.to(dtype=dtype, device=device)
+        inpt = torch.rand(2, 5, dtype=dtype, device=device)
+
+        optim_inputs = optim_info.optim_inputs_func(device=device)
+        
+        mem_used_exp_avg = 0 
+        case_to_mem_usage = {}
+        
+        def get_obj_size(d):
+            if isinstance(d, dict):
+                return sum(get_obj_size(v) for v in d.values())
+            elif isinstance(d, (list,tuple)):
+                return sum(get_obj_size(v) for v in d)
+            elif isinstance(d, torch.Tensor):
+                return d.element_size() * d.numel()
+            return d.__sizeof__()
+            
+        for optim_input in optim_inputs:
+            beta1_zero = (
+                "betas" in optim_input.kwargs and optim_input.kwargs["betas"][0] == 0.0
+            )
+            if beta1_zero or optim_input.desc == "default":
+                optim = optim_info.optim_cls(model.parameters(), **optim_input.kwargs)
+                optim.zero_grad() 
+                output = model(inpt)
+                loss = output.sum()
+                loss.backward()
+                optim.step()
+
+                if not beta1_zero : 
+                    mem_used_exp_avg += sum(
+                        optim.state[p]["exp_avg"].element_size()
+                        * optim.state[p]["exp_avg"].numel()
+                        for p in model.parameters()
+                    ) 
+                    
+                case_to_mem_usage[
+                    "beta1_zero" if beta1_zero else "default"
+                ] = get_obj_size(optim.state_dict()) 
+            
+                
+        self.assertGreaterEqual(
+            case_to_mem_usage["default"]-case_to_mem_usage["beta1_zero"], mem_used_exp_avg 
+        )
+        
+
+    @optims(
+        [optim for optim in optim_db if optim.optim_cls.__name__ == "Adam"],
+        dtypes=[torch.float32],
+    )
+    def test_loadstate_with_diff_beta1_adam(
+        self, device, dtype, optim_info
+    ):
+        # the test takes form of two tests. 
```

Comment:
What exactly is this test testing? It looks like it tests that loading a beta1=0 optim into a default optim will compute similar results, which doesn't have to be true, right?

Could you see if you could simplify the test more to make it more readable?

One way to do so:
1. create two optimizers, one with beta1 is zero, the other nonzero
2. run both for a few steps
3. switch their state!
4. run both again for a few steps

There's no need to assert equality, as they're not expected to be equal to each other, just that they can be loaded into each other!

Reply:
If I understand the test currently written, then it is doing 

first N steps:
params A get updates with a nonzero beta
params B get updates with zero beta

then second N steps:
params A get updates with zero beta (since it got swapped)
params B get updates with nonzero beta

which I don't think is equivalent. I DO think the following is equivalent:

first N steps:
params A get updates with a nonzero beta

then we load a state dict into existing optim A and a new optim B both now with beta

then we run for more steps and check that those are the same.

Reply:
> first N steps:
params A get updates with a nonzero beta
params B get updates with zero beta
then second N steps:
params A get updates with zero beta (since it got swapped)
params B get updates with nonzero beta

I do not do any swapping. I do something more like what you explained after. More exactly : 

First N steps : 
params A get updates with a nonzero beta
params B : nothing 

Load A with a state that has beta = 0. 

Second N steps: 
params A get updates with zero beta.  
params B get updates with zero beta (its initial beta). 

At the end I expect A and B to behave the same.

Reply:
ah but the params in A and B need to be matching between the first and second N steps, right?